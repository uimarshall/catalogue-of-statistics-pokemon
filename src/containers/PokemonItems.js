import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import Spinner from 'react-bootstrap/Spinner';
import { fetchPokemons } from '../actions/pokemonActions';
import { getPokemonsError, getAllPokemons, getPokemonsPending } from '../reducers/pokemonsReducer';
import { getPokemonType } from '../reducers/filterReducer';
import DisplayPokemon from '../components/DisplayPokemon';
import PokemonFilter from '../components/PokemonFilter';
import AlertDismissible from '../utils/Alert';

class PokemonItems extends React.Component {
  constructor(props) {
    super(props);
    this.handleFilterChange = this.handleFilterChange.bind(this);
  }

  componentDidMount() {
    const { fetchPokemons } = this.props;
    fetchPokemons('normal');
  }

  handleFilterChange(e) {
    const { fetchPokemons } = this.props;
    if (e.target.value !== '') {
      fetchPokemons(e.target.value);
    }
    e.preventDefault();
  }

  render() {
    const { data, filter } = this.props;
    const { error, pending, pokemons } = data;
    if (error) {
      <AlertDismissible />;

      return (
        <div className="error">

          <AlertDismissible />

        </div>
      );
    }

    if (pending) {
      return (
        <div className="d-flex justify-content-center">
          <Spinner animation="grow" />
        </div>
      );
    }

    if (pokemons.length < 2) {
      return (
        <div className="d-flex justify-content-center">
          <Spinner animation="grow" />
        </div>
      );
    }

    return (
      <div className="pokemon-item mx-auto">
        <PokemonFilter onClick={this.handleFilterChange} category={filter} />
        <ul className="d-flex container mx-auto my-2 flex flex-wrap">
          <div className="display-pokemon row">
            {pokemons.map(pokemon => (
              <DisplayPokemon
                key={pokemon.name}
                pokemon={pokemon}
              />
            ))}
          </div>
        </ul>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchPokemons,
}, dispatch);

const mapStateToProps = state => ({
  data: {
    error: getPokemonsError(state.data),
    pokemons: getAllPokemons(state.data),
    pending: getPokemonsPending(state.data),
  },
  filter: getPokemonType(state.filter),
});

PokemonItems.defaultProps = {
  data: {
    pending: true,
    error: null,
    pokemons: [],
  },
  filter: 'normal',
};

PokemonItems.propTypes = {
  data: PropTypes.shape({
    pending: PropTypes.bool,
    error: PropTypes.objectOf(PropTypes.string),
    pokemons: PropTypes.arrayOf(PropTypes.object),
  }),
  filter: PropTypes.string,
  fetchPokemons: PropTypes.func.isRequired,
};
export default connect(mapStateToProps, mapDispatchToProps)(PokemonItems);
