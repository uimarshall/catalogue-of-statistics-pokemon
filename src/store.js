/* eslint-disable no-underscore-dangle */
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducers';

const initialState = {
  pending: true,
  pendingPokemon: true,
  pokemons: [],
  error: null,
};

const middleware = [thunk];
const devtools = process.env.NODE_ENV === 'test'
  ? x => x
  : window.__REDUX_DEVTOOLS_EXTENSION__
      && window.__REDUX_DEVTOOLS_EXTENSION__();

const store = createStore(
  rootReducer,
  {
    data: initialState,
    filter: 'normal',
  },
  compose(
    applyMiddleware(...middleware),
    devtools,
  ),
);

export default store;
