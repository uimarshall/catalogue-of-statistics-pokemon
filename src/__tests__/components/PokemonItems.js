import '@testing-library/jest-dom';
import * as React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import store from '../../store';
import PokemonItems from '../../containers/PokemonItems';

describe('Proper Rendering', () => {
  test('renders PokemonItems component correctly', () => {
    render(<Provider store={store}><PokemonItems /></Provider>);

    expect(screen.queryByText(/Run faster now/)).toBeNull();
  });
});

describe('PokemonItems Component', () => {
  afterEach(cleanup);

  it('should take a snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}><Router><PokemonItems /></Router></Provider>,
    );

    expect(
      asFragment(<Provider store={store}><Router><PokemonItems /></Router></Provider>),
    ).toMatchSnapshot();
  });
});
