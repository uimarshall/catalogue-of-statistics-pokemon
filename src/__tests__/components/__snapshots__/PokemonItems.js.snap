// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonItems Component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center align-items-center"
  >
    <div
      class="spinner-border text-warning"
    />
  </div>
</DocumentFragment>
`;
